### PR TITLE
[JENKINS-71728] Add CSP compatible version of `FormApply#applyResponse`

### PR DESCRIPTION
**Pull request by yaroslavafenkin**
See JENKINS-71728.

### Testing done

* JENKINS-71728: tested by clicking the Apply button on various configuration screens and making sure no errors show up in developer console. There's also a caller in [Descriptor.java#L1192](https://github.com/jenkinsci/jenkins/blob/0e5f23813fa735b444717fd1ed7a83eb56822b15/core/src/main/java/hudson/model/Descriptor.java#L1192) but I wasn't able to invoke that. Checking callers I see that they're deprecated. I assume that piece exists for compatibility reasons? If not and there are actual usages that I've missed, let me know and I'd be happy to test that.

### Proposed changelog entries

N/A

### Proposed upgrade guidelines

N/A

```[tasklist]
### Submitter checklist
- [x] The Jira issue, if it exists, is well-described.
- [x] The changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developers, depending on the change) and are in the imperative mood (see [examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)). Fill in the **Proposed upgrade guidelines** section only if there are breaking changes or changes that may require extra steps from users during upgrade.
- [ ] There is automated testing or an explanation as to why this change has no tests.
- [ ] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadocs, as appropriate.
- [ ] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")`, if applicable.
- [x] New or substantially changed JavaScript is not defined inline and does not call `eval` to ease future introduction of Content Security Policy (CSP) directives (see [documentation](https://www.jenkins.io/doc/developer/security/csp/)).
- [ ] For dependency updates, there are links to external changelogs and, if possible, full differentials.
- [ ] For new APIs and extension points, there is a link to at least one consumer.
```

### Desired reviewers

@jenkinsci/core-security-review

Before the changes are marked as `ready-for-merge`:

```[tasklist]
### Maintainer checklist
- [ ] There are at least two (2) approvals for the pull request and no outstanding requests for change.
- [ ] Conversations in the pull request are over, or it is explicit that a reviewer is not blocking the change.
- [ ] Changelog entries in the pull request title and/or **Proposed changelog entries** are accurate, human-readable, and in the imperative mood.
- [ ] Proper changelog labels are set so that the changelog can be generated automatically.
- [ ] If the change needs additional upgrade steps from users, the `upgrade-guide-needed` label is set and there is a **Proposed upgrade guidelines** section in the pull request title (see [example](https://github.com/jenkinsci/jenkins/pull/4387)).
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).
```

<a href="https://gitpod.io/#https://github.com/jenkinsci/jenkins/pull/8351"></a>

### CODE REVIEW

**Comment by daniel-beck** [comment]
Re CodeMirror, please note I my ongoing attempt of that in #6867. Lots of plugins set invalid JSON as parameter which would result in this breaking.

**Comment by yaroslavafenkin** [comment]
> Re CodeMirror, please note I my ongoing attempt of that in https://github.com/jenkinsci/jenkins/pull/6867. Lots of plugins set invalid JSON as parameter which would result in this breaking.

Oh that slipped. I guess I'm better off just reverting that part.

**Review by daniel-beck — CHANGES_REQUESTED** [review]
Latest changes appear to be breaking: https://github.com/search?q=org%3Ajenkinsci+FormApply+applyResponse&type=code

**Review by daniel-beck — COMMENTED** [review]
A good solution for deprecation of `FormApply#applyResponse` will provide an alternative API for known outside uses.

Most appear compatible with this change (i.e., can easily be migrated), but https://github.com/jenkinsci/leanix-microservice-intelligence-plugin/blob/be5ba1f015bd72131620d10e315cca8eda7c77cb/src/main/java/io/jenkins/plugins/leanixmi/SettingsPanel.java#L56 and https://github.com/jenkinsci/credentials-plugin/blob/54b1c2c6388a3eaf49d3bd4ab55ab15b2abc2482/src/main/java/com/cloudbees/plugins/credentials/CredentialsSelectHelper.java#L602 (from a quick GH search) look like they cannot be adapted to the new API.

Perhaps plugins could put the JS into a resource file, and specify its file name as parameter, with the generated HTML including that JS file?

Alternatively, explain why the known incompatible uses are a bad use of this API (and ideally, what the plugins should do instead).

**Comment by yaroslavafenkin** [comment]
> Most appear compatible with this change (i.e., can easily be migrated), but https://github.com/jenkinsci/leanix-microservice-intelligence-plugin/blob/be5ba1f015bd72131620d10e315cca8eda7c77cb/src/main/java/io/jenkins/plugins/leanixmi/SettingsPanel.java#L56 and https://github.com/jenkinsci/credentials-plugin/blob/54b1c2c6388a3eaf49d3bd4ab55ab15b2abc2482/src/main/java/com/cloudbees/plugins/credentials/CredentialsSelectHelper.java#L602 (from a quick GH search) look like they cannot be adapted to the new API.

https://plugins.jenkins.io/leanix-microservice-intelligence/ is now deprecated, I don't think it'll ever get jenkins baseline version including this change.
For credentials-plugin suggestion see https://github.com/jenkinsci/credentials-plugin/pull/481.

After spending some time looking into the re-work for credentials plugin I think that the way to go would be to completely get rid of `FormApply#applyResponse` (or rather deprecate it), as the whole thing seems like a hack to me now. Instead just send data as json in a response, and let javascript use the data to display the correct notification.

**Review by Kevin-CB — APPROVED** [review]
Well, I tested this PR locally with the CSP plugin and it works as expected. However, I also tried it without the fix and got no CSP violations with the plugin.

I'm now curious as to why this behavior doesn't violate the configured CSP.

File: core/src/main/java/hudson/util/FormApply.java
Comment (daniel-beck) [inline core/src/main/java/hudson/util/FormApply.java:108]: Callback seems like a confusing term, because it doesn't seem to apply here.

File: core/src/main/java/hudson/util/FormApply.java
Comment (daniel-beck) [inline core/src/main/java/hudson/util/FormApply.java:100]: Probably a good idea to reference the UI documentation that specifies the notification types.

**Comment by daniel-beck** [comment]
> I'm now curious as to why this behavior doesn't violate the configured CSP.

`#generateResponse` does not set a CSP header, and its output is not `l:layout` based, so no `PageDecorator` is applied. This change makes the former trivial to do though, and similar to other occurrences of CSP fixes, the fix alone, removing inline JS, is helpful.

File: core/src/main/java/hudson/util/FormApply.java
Comment (yaroslavafenkin) [reply core/src/main/java/hudson/util/FormApply.java:100]: I guess the best we can do is link to https://github.com/jenkinsci/jenkins/blob/74610e024a6b8fd8feccdc51b8f7741aa6c30e3b/war/src/main/js/components/notifications/index.js#L13-L25 Would that be good enough?

Comment (daniel-beck) [reply core/src/main/java/hudson/util/FormApply.java:100]: Awkward that there's no documentation for this, so probably the best we can do.

File: core/src/main/java/hudson/util/FormApply.java
Comment (daniel-beck) [inline core/src/main/java/hudson/util/FormApply.java:106]: What motivated the removal of `<html><body>…</body></html>` from the response?

File: core/src/main/java/hudson/util/FormApply.java
Comment (daniel-beck) [inline core/src/main/java/hudson/util/FormApply.java:101]: I wonder whether the second arg might be better off as an `enum` defined in this class. Then there's no need to `#htmlAttributeEscape` for lack of ability to put something else there.

**Review by daniel-beck — APPROVED** [review]

File: core/src/main/java/hudson/util/FormApply.java
Comment (yaroslavafenkin) [reply core/src/main/java/hudson/util/FormApply.java:101]: Thought about that too, will add.

File: core/src/main/java/hudson/util/FormApply.java
Comment (yaroslavafenkin) [reply core/src/main/java/hudson/util/FormApply.java:106]: I have no good reason. It's just that it works without `<html><body>…</body></html>`. Those don't make the HTML returned a complete one.

**Review by daniel-beck — APPROVED** [review]

File: core/src/main/java/hudson/util/FormApply.java
Comment (NotMyFault) [inline core/src/main/java/hudson/util/FormApply.java:78]:
```suggestion
     * @deprecated use {@link #showNotification(String, NotificationType)} instead, which is CSP compatible version
```
Otherwise, it'll render as `use {#showNotification(String, FormApply.NotificationType)}...` instead of `use showNotification(String, FormApply.NotificationType)...`

**Review by NotMyFault — APPROVED** [review]

File: core/src/main/java/hudson/util/FormApply.java
Comment (yaroslavafenkin) [reply core/src/main/java/hudson/util/FormApply.java]: Thanks for the catch, addressed in https://github.com/jenkinsci/jenkins/pull/8351/commits/e8716aee22a04b475d0c6a499b186ca2f01d6c75

File: core/src/main/java/hudson/util/FormApply.java
Comment (daniel-beck) [inline core/src/main/java/hudson/util/FormApply.java:99]:
```suggestion
     * that will show a notification of certain type and with provided message.
     * @since TODO
```

**Comment by NotMyFault** [comment]
/label ready-for-merge

---

This PR is now ready for merge. We will merge it after ~24 hours if there is no negative feedback.
Please see the [merge process documentation](https://github.com/jenkinsci/jenkins/blob/master/docs/MAINTAINERS.adoc#merge-process) for more information about the merge process.
Thanks!

File: core/src/main/java/hudson/model/Descriptor.java
Comment (daniel-beck) [inline core/src/main/java/hudson/model/Descriptor.java:1201]: This looks wrong. The `#quote` call seems to have been needed because this built a JavaScript string and needed to escape e.g. `"`. That's no longer the case, so the call should be removed.

File: core/src/main/java/hudson/util/FormApply.java
Comment (daniel-beck) [inline core/src/main/java/hudson/util/FormApply.java:102]: @yaroslavafenkin Make it clear through naming and/or Javadoc what the HTML content of `message` is. The call looks like it would support HTML, but needs to be documented.

**Review by daniel-beck — CHANGES_REQUESTED** [review]

File: core/src/main/java/hudson/util/FormApply.java
Comment (yaroslavafenkin) [reply core/src/main/java/hudson/util/FormApply.java:102]: What do you mean by the HTML content of `message`? I'm not catching it.

Comment (daniel-beck) [reply core/src/main/java/hudson/util/FormApply.java:102]: Is `message` plain text/escaped, or does it support HTML? There's no Javadoc for the argument, so this is undetermined.

**Review by daniel-beck — APPROVED** [review]

**Comment by basil** [comment]
What are the next steps for this PR?

**Comment by basil** [comment]
@yaroslavafenkin Any remaining action items here?

**Comment by yaroslavafenkin** [comment]
> @yaroslavafenkin Any remaining action items here?

I can't recall any, but it's been a while. I'll retest this next week and give it a green light if everything is still alright.

**Comment by yaroslavafenkin** [comment]
Retested, seems OK to me. Interestingly this doesn't get caught by [csp plugin](https://plugins.jenkins.io/csp/).

**Comment by timja** [comment]
/label ready-for-merge

----

This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

**Pull request merged**
